### DISCUSSION TITLE
AD

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#22257] [Q&A General] [answered]
Hello

Just a quick question from your workshop.
In the following Logan wrote `**_pressure_gradient(coupledGradient("pressure")),**` in the DarcyVelocity AuxKernel file.
I can see material is with AD, why do not we have ad for pressure such as `**_pressure_gradient(adCoupledGradient("pressure")),**`
In the opposite way, can we remove the AD from the _permeability(getADMaterialProperty<Real>("permeability")), etc?
Thanks.

-------------------------------------------------------------------------------------
DarcyVelocity::DarcyVelocity(const InputParameters & parameters)
: VectorAuxKernel(parameters),
// Get the gradient of the variable
`**_pressure_gradient(coupledGradient("pressure")),**`
// Set reference to the permeability MaterialProperty.
// Only AuxKernels operating on Elemental Auxiliary Variables can do this
_permeability(getADMaterialProperty<Real>("permeability")),
// Set reference to the viscosity MaterialProperty.
// Only AuxKernels operating on Elemental Auxiliary Variables can do this
_viscosity(getADMaterialProperty<Real>("viscosity"))
{}
RealVectorValue
DarcyVelocity::computeValue()
{
// Access the gradient of the pressure at this quadrature point, then pull out the
"component" of
// it requested (x, y or z). Note, that getting a particular component of a
gradient is done using
// the parenthesis operator.
return -MetaPhysicL::raw_value(_permeability[_qp] / _viscosity[_qp]) *
_pressure_gradient[_qp];

**Comment by Oops-Qiao** [comment, marked as answer]
Just another question here again.

pressure is a real value and velocity is vector value, right?
the following statement is correct or not?

  _p(adCoupledValue("pressure")),
  _grad_p(adCoupledGradient("pressure"))
After taking the gradient of pressure, _grad_p will become as a vector value, right?

then if I treat gradient of pressure as the coupled value, it will become as follows, am I right?

  _grad_p(adCoupled**Vector**Value("_grad_p"))

Thanks again.

**Reply by GiudGiud** [reply]
No you can’t do that because grad_p is not a variable name. 

ADCoupledGradient is getting you the gradient of a variable, and storing it as a vector

**Reply by Oops-Qiao** [reply]
Thanks a lot, I see your point.
